All Gather and gather APIs for Python Objects

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42189 All Gather and gather APIs for Python Objects**

Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

rohan-varma committed:
All Gather and gather APIs for Python Objects 

Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1156,6 +1157,148 @@ def all_gather_multigpu(output_tensor_lists,
+        group (ProcessGroup, optional): The process group to work on
+
+    Returns:
+        None. If the calling rank is part of this group, the output of the
```
Let's also say that if the rank is not in the group, the `object_list` argument will stay intact.

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1156,6 +1157,148 @@ def all_gather_multigpu(output_tensor_lists,
+    # Avoid copying intermediate tensors back and forth to CUDA by using gloo PG
+    # for object collectives.
+    if get_backend(group) == Backend.NCCL:
+        gloo_group = new_group(
```
this will introduce one more rendezvous, will this be faster than copy the tensor to GPU when using NCCL backend?

rohan-varma replied:
I see, maybe we can just go with the transfer to CUDA approach. I was thinking if we can cache this gloo group, then we might pay less amortized cost if this function is called many times.

mrshenli replied:
I see, let's use the provided group in this PR. If we got complain regarding its speed, we can find ways to optimize. Another reason for not creating a new Gloo group is that users might build `distributed` package without Glo.

rohan-varma replied:
Sounds good, I think this can work for `all_gather_object` but for `gather_object`, we get the error `RuntimeError: ProcessGroupNCCL does not support gather`. I guess if users try to call `all_gather_object` with NCCL backend, it makes sense to throw, since NCCL does not implement gather. So should we just throw in this case?

mrshenli replied:
I see. Yep, I think it makes sense to just throw when calling `gather_object` with NCCL backend.

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1156,6 +1157,148 @@ def all_gather_multigpu(output_tensor_lists,
+    local_max_size_tensor = torch.ByteTensor(size=(max_object_size,))
+    local_max_size_tensor[: local_size.item()] = input_tensor
+    output_tensors = [
+        torch.empty(max_object_size, dtype=torch.uint8) for _ in range(group_size)
```
(not sure if this can work) we can avoid multiple memory allocation by creating a big tensor and then let output tensors points to its views.

rohan-varma replied:
We would still end up allocating the same amount of memory right? Although there may be a perf win by doing less memory allocation calls.

mrshenli replied:
Yep, exactly.

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1156,6 +1157,148 @@ def all_gather_multigpu(output_tensor_lists,
+def all_gather_object(object_list, obj, group=group.WORLD):
+    """
+    Gathers picklable objects from the whole group into a list. Similar to
+    all_gather, but Python objects can be passed in. Note that the object must
```
Shall we make `all_gather` link to the function?

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1156,6 +1157,148 @@ def all_gather_multigpu(output_tensor_lists,
+def gather_object(obj, object_gather_list=None, dst=0, group=group.WORLD):
+    """
+    Gathers picklable objects from the whole group in a single process.
+    Similar to gather, but Python objects can be passed in. Note that the object
```
Shall we make `gather` link to the function?

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1156,6 +1157,148 @@ def all_gather_multigpu(output_tensor_lists,
+
+    Arguments:
+        obj (Any): Input object. Must be picklable.
+        object_gather_list (list[Any]): Output list. On the dst rank, it should
```
double quote dst

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1156,6 +1157,148 @@ def all_gather_multigpu(output_tensor_lists,
+        obj (Any): Input object. Must be picklable.
+        object_gather_list (list[Any]): Output list. On the dst rank, it should
+        be correctly sized as the size of the group for this collective and will
+        contain the output. Must ne None on non-dst ranks. Default=None.
```
ne -> be

mrshenli replied:
Let's make the Default statement consistent with the the one below and `gather`. (Default is None)

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1156,6 +1157,148 @@ def all_gather_multigpu(output_tensor_lists,
 
+def _object_to_tensor(obj):
+    buffer = io.BytesIO()
+    torch.save(obj, buffer)
```
I see, I went with `torch.save`/`torch.load` since it seems that is the de-facto way of doing serialization in PT. I think it might be more efficient to just use the `pickle` module directly, though.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1156,6 +1157,148 @@ def all_gather_multigpu(output_tensor_lists,
         work.wait()
 
 
+def _object_to_tensor(obj):
```
I didn't see a need to add these the protocols since we have control over most of the serialization here, i.e. we directly serialize and deserialize ourselves. Looking at the [docs](https://docs.python.org/3/library/pickle.html) it seems that specifying a higher protocol might give some performance win over a lower protocol, but I would advocate for keeping this interface simpler if we don't see demand for it.

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1156,6 +1157,169 @@ def all_gather_multigpu(output_tensor_lists,
+        since it does not provide an async_op handle and thus will be a blocking
+        call.
+
+    ..note:: Note that this API is not supported when using the NCCL backend.
```
missing a space between ".." and "note"?

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1156,6 +1157,169 @@ def all_gather_multigpu(output_tensor_lists,
+        group: (ProcessGroup, optional): The process group to work on.
+
+    Returns:
+        None. On the dst rank, object_gather_list will contain the output of the
```
double quote code?

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

rohan-varma commented:
Failure is unrelated: 

```
Aug 05 18:38:10 ======================================================================
Aug 05 18:38:10 FAIL: test_unary_ops (__main__.TestTensorExprFuser)
Aug 05 18:38:10 ----------------------------------------------------------------------
Aug 05 18:38:10 Traceback (most recent call last):
Aug 05 18:38:10   File "test_tensorexpr.py", line 901, in test_unary_ops
Aug 05 18:38:10     np.testing.assert_allclose(x.cpu().numpy(), y.cpu().numpy())
Aug 05 18:38:10   File "/opt/conda/lib/python3.6/site-packages/numpy/testing/_private/utils.py", line 1528, in assert_allclose
Aug 05 18:38:10     verbose=verbose, header=header, equal_nan=equal_nan)
Aug 05 18:38:10   File "/opt/conda/lib/python3.6/site-packages/numpy/testing/_private/utils.py", line 764, in assert_array_compare
Aug 05 18:38:10     flagged = func_assert_same_pos(x, y, func=isnan, hasval='nan')
Aug 05 18:38:10   File "/opt/conda/lib/python3.6/site-packages/numpy/testing/_private/utils.py", line 740, in func_assert_same_pos
Aug 05 18:38:10     raise AssertionError(msg)
Aug 05 18:38:10 AssertionError: 
Aug 05 18:38:10 Not equal to tolerance rtol=1e-07, atol=0
Aug 05 18:38:10 
Aug 05 18:38:10 x and y nan location mismatch:
Aug 05 18:38:10  x: array([1., 1., 1., ..., 1., 1., 1.], dtype=float32)
Aug 05 18:38:10  y: array([nan, nan, nan, ..., nan, nan, nan], dtype=float32)
Aug 05 18:38:10 
Aug 05 18:38:10 ----------------------------------------------------------------------
```

rohan-varma committed:
Update on "All Gather and gather APIs for Python Objects "


Rehash of https://github.com/pytorch/pytorch/pull/28811, which was several months old.

As part of addressing https://github.com/pytorch/pytorch/issues/23232, this PR adds support for the following APIs:

`allgather_object` and `gather_object` to support gather/allgather of generic, pickable Python objects. This has been a long-requested feature so PyTorch should provide these helpers built-in.

The methodology is what is proposed in the original issue:
1) Pickle object to ByteTensor using torch.save
2) Comm. tensor sizes
3) Copy local ByteTensor into a tensor of maximal size
4) Call tensor-based collectives on the result of (3)
5) Unpickle back into object using torch.load

Note that the API is designed to match the tensor-based collectives other than supporting `async_op`. For now, it is a blocking call. If we see demand to support `async_op`, we will have to make more progress on merging work/future to support this.

If this is a suitable approach, we can support `scatter`, `broadcast` in follow up PRs.

Differential Revision: [D22785387](https://our.internmc.facebook.com/intern/diff/D22785387/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22785387/)!

[ghstack-poisoned]

kcmao commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1156,6 +1157,169 @@ def all_gather_multigpu(output_tensor_lists,
+
+
+def _tensor_to_object(tensor, tensor_size):
+    buf = tensor.numpy().tobytes()[:tensor_size]
```
csd
